(CDPE-3239) Add functions for pipeline reporting

kreeuwijk opened the pull request:
This PR adds new functions to enable dynamic reporting of the pipeline while it's running:
- `cd4pe_deployments::get_impact_analysis` adds the ability to get the environments analyzed by IA
- `cd4pe_deployments::get_pipeline` adds the ability to get the last run result for a specific pipeline
- `cd4pe_deployments::get_pipeline_trigger_event` adds the ability to get the run result for a recently run pipeline, by pipeline ID and commit SHA
- `cd4pe_deployments::search_impacted_nodes` adds the ability to get the detailed impact on individual nodes from an IA
- `cd4pe_deployments::search_pipeline` adds the ability to get the pipeline ID for a pipeline that was triggered on a specific commit SHA
- `cd4pe_deployments::get_cookie` adds the ability to generate a CD4PE cookie from basic auth, to interact with API operations that do not allow token auth at this point (even though the added functions do all use token auth now)

This PR contains updates to `cd4pe_client.rb` to allow all the above functions to work properly. It also contains some minor rubocop fixes.

eputnam commented on lib/puppet/functions/cd4pe_deployments/get_cookie.rb:
```diff
@@ -0,0 +1,36 @@
+
+    response = client.get_cookie(login_user, login_pwd)
+    case response
+    when Net::HTTPSuccess
```
these Net:HTTPx classes should work ok in this context. I think one reason we match codes instead is because there's a possibility that CD4PE returns a 302 from these Ajax endpoints in the case that the user is not logged in. That being the case, these _should_ be ok and it certainly looks nicer.

kreeuwijk commented:
@eputnam can you perform the merge? I don't have permissions for that.
I've updated REFERENCE.md and fixed some inconsistencies that lingered in the puppet strings documentation. Also added a generic evaluate_result helper function.

eputnam commented:
@kreeuwijk , this looks awesome! before we merge, I have a couple git requests. 
1. squash f0a21ab and e33c1fa together
2. reset back to e9abca9 and do a rebase on master instead of a merge
